hive-bigquery-connector-common/src/main/java/com/google/cloud/hive/bigquery/connector/config/HiveBigQueryConnectorUserAgentProvider.java

```diff
@@ -16,17 +16,84 @@
 package com.google.cloud.hive.bigquery.connector.config;
 
 import com.google.cloud.bigquery.connector.common.UserAgentProvider;
+import com.google.cloud.hive.bigquery.connector.utils.MiscUtils;
+import com.google.common.io.CharStreams;
+import com.google.common.io.Closeables;
+import com.google.gson.JsonObject;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.nio.charset.StandardCharsets;
+import java.util.Optional;
+import org.apache.http.client.config.RequestConfig;
+import org.apache.http.client.methods.CloseableHttpResponse;
+import org.apache.http.client.methods.HttpGet;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.impl.client.HttpClients;
 
 public class HiveBigQueryConnectorUserAgentProvider implements UserAgentProvider {
 
-  private final String hiveQueryID;
+  private Optional<String> gpn;
+  static String GCP_REGION_PART = getGcpRegion().map(region -> " region/" + region).orElse("");
+  private static String JAVA_VERSION = System.getProperty("java.runtime.version");
+  static String DATAPROC_IMAGE_PART =
+      Optional.ofNullable(System.getenv("DATAPROC_IMAGE_VERSION"))
+          .map(image -> " dataproc-image/" + image)
+          .orElse("");
 
-  public HiveBigQueryConnectorUserAgentProvider(String hiveQueryID) {
-    this.hiveQueryID = hiveQueryID;
+  public HiveBigQueryConnectorUserAgentProvider(Optional<String> gpn) {
+    this.gpn = gpn;
+  }
+
+  static Optional<String> getGcpRegion() {
+    RequestConfig config =
+        RequestConfig.custom()
+            .setConnectTimeout(100)
+            .setConnectionRequestTimeout(100)
+            .setSocketTimeout(100)
+            .build();
+    CloseableHttpClient httpClient = HttpClients.custom().setDefaultRequestConfig(config).build();
+    HttpGet httpGet =
+        new HttpGet("http://metadata.google.internal/computeMetadata/v1/instance/zone");
+    httpGet.addHeader("Metadata-Flavor", "Google");
+    try (CloseableHttpResponse response = httpClient.execute(httpGet)) {
+      if (response.getStatusLine().getStatusCode() == 200) {
+        String body =
+            CharStreams.toString(
+                new InputStreamReader(response.getEntity().getContent(), StandardCharsets.UTF_8));
+        return Optional.of(body.substring(body.lastIndexOf('/') + 1));
+      } else {
+        return Optional.empty();
+      }
+    } catch (Exception e) {
+      return Optional.empty();
+    } finally {
+      try {
+        Closeables.close(httpClient, true);
+      } catch (IOException e) {
+        // nothing to do
+      }
+    }
   }
 
   @Override
   public String getUserAgent() {
-    return "Hive-Query-" + hiveQueryID;
+    StringBuilder userAgentBuilder = new StringBuilder();
+    userAgentBuilder.append("hive-bigquery-connector/").append(MiscUtils.CONNECTOR_VERSION);
```

Comment:
why do we need this connector-version in the userAgent? it looks a lot extra work to add a resource file and a MiscUtils for it

Reply:
Another reason I can think of is that this would allow us to track what versions are effectively used out in the wild.

Reply:
can this be moved to JobUtils instead of a new MiscUtils?

Reply:
Sure!

Reply:
@yigress  Ok I've moved the code to JobUtils and removed MiscUtils.